tests/test_validation.py

```diff
@@ -1260,3 +1323,75 @@ def test_allow_extra_keys(self):
         with self.assertLogs('natcap.invest.validation', level='DEBUG') as cm:
             validation.validate(args, spec)
         self.assertTrue(message in cm.output)
+
+    def test_check_ratio(self):
+        """Validation: test ratio type validation."""
+        from natcap.invest import validation
+        args = {
+            'a': 'xyz',  # not a number
+            'b': '1.5',  # too large
+            'c': '-1',   # too small
+            'd': '0',    # lower bound
+            'e': '0.5',  # middle
+            'f': '1'     # upper bound
+        }
+        spec = {name: {'type': 'ratio'} for name in args}
+
+        expected_warnings = [
+            (['a'], 'Value "xyz" could not be interpreted as a number'),
+            (['b'], 'Value 1.5 is not in the range [0, 1]'),
+            (['c'], 'Value -1.0 is not in the range [0, 1]')]
+        actual_warnings = validation.validate(args, spec)
+        for warning in actual_warnings:
+            self.assertTrue(warning in expected_warnings)
+
+    def test_check_percent(self):
+        """Validation: test percent type validation."""
+        from natcap.invest import validation
+        args = {
+            'a': 'xyz',    # not a number
+            'b': '100.5',  # too large
+            'c': '-1',     # too small
+            'd': '0',      # lower bound
+            'e': '55.5',   # middle
+            'f': '100'     # upper bound
+        }
+        spec = {name: {'type': 'percent'} for name in args}
+
+        expected_warnings = [
+            (['a'], 'Value "xyz" could not be interpreted as a number'),
+            (['b'], 'Value 100.5 is not in the range [0, 100]'),
+            (['c'], 'Value -1.0 is not in the range [0, 100]')]
+        actual_warnings = validation.validate(args, spec)
+        for warning in actual_warnings:
+            self.assertTrue(warning in expected_warnings)
+
+    def test_check_code(self):
+        """Validation: test code type validation."""
+        from natcap.invest import validation
+        args = {
+            'a': 'xyz',    # not a number
+            'b': '1.5',    # not an integer
+            'c': '-1',     # too small
```

Comment:
I think I commented this somewhere else. I wonder if negative values should be restricted from codes. They do come up sometimes in LULC rasters biophysical tables.

Reply:
Good point. Changed this to allow negative integers in 0ebd1746